### PR TITLE
initial merkle tree implementation

### PR DESCRIPTION
**Pull request by Warchant**
- can calculate merkle root
- can calculate the tree
- can not provide merkle proof (a path)

### CODE REVIEW

File: src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+  public MerkleTree(@NonNull MessageDigest digest) {
+    this.digest = digest;
+  }
+
+  public byte[] root() {
+    return tree[0];
+  }
+
+  public void create(List<byte[]> level) {
```
[the remark this answers is not in the document]
Comment (JIexa) [reply src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java]: do we need JavaDoc for this class?

Comment (JIexa) [inline src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java]: also, it is not clear what level is. 
Intuitively, the level should be smth like a numeric value or string, but not a collection.

File: src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  // [root] [intermediate nodes] [leafs]
+  private byte[][] tree;
+
+  /**
+   * Creates new tree from given leafs with given digest function.
+   * @param digest implementation of the digest algorithm
+   * @param leafs list of leaf nodes
+   */
+  public MerkleTree(@NonNull MessageDigest digest, @NonNull List<byte[]> leafs) {
```
Comment (tdorzhi) [inline src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java]: Having two constructors and two flows of working with this object seems confusing
Caller can either call constructor and pass `leafs` immediately, or just create object with `digest` only and then use `#create`. Leave only one way of using, and get rid of other.

Comment (tdorzhi) [reply src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java]: I recommend leaving second one, when object is created with `digest` only.
Thus we separate initialization from behaviour

File: src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+   */
+  public byte[] root() {
+    return tree[0];
+  }
+
+  /**
+   * Calculates merkle tree given leafs.
+   * @param level the bottom-most level of the tree (e.g. leafs)
+   */
+  public void create(List<byte[]> level) {
```
Comment (tdorzhi) [inline src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java]: `MerkleTree` should be separated to two classes, one which *transforms* `List<byte[]>` to Merkle tree, and one which *represents* Merkle tree
I suggest renaming this class to LeafsToMerkleTree and changing return value of method `#create` to `MerkleTree`, which will be new class which only represents Merkle tree

File: src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  // [root] [intermediate nodes] [leafs]
+  byte[][] tree;
+
+  /**
+   * Getter for the tree. Tree is stored "by levels", so first level is root and stored
+   * as tree[0], next level (2 items) is stored in tree[1] and tree[2], and so on.
+   * @return tree represented as [root] [intermediate nodes] [leafs]
+   */
+  public byte[][] getTree() {
```
Comment (tdorzhi) [inline src/main/java/jp/co/soramitsu/sora/crypto/MerkleTree.java]: remove this method, as `@Value` generates getters automatically, move java docs to the field

**Pull request merged**